### ISSUE TITLE
Freeze / unreactive app on Windows

### ISSUE THREAD

**Issue by biva** [mattermost/desktop#1851]
**Summary**
The desktop application is not reactive anymore: I click or write something, but nothing happens.

**Environment**
- Operating System: Windows 10
- Mattermost Desktop App version: 5.0.1
- Mattermost Server version: 6.0.2

**Steps to reproduce**
I'm not sure exactly when it happens, but I have the feeling that it happens when the app is reduced in the tray (see picture below)
![image](https://user-images.githubusercontent.com/9964305/140301209-69132720-305e-4103-8326-79e534d6e94b.png)
1. Click on the icon to open it
2. Click on a channel (for example)

**Expected behavior**
3. The channel should open

**Observed behavior**
4. Nothing happens during several seconds, then the app is reactive again and I can use it normally... until the next time the bug occurs

This never happened before, when I was using server 5.39 and the previous desktop app (I'm not sure which version it was, sorry, but probably v4.7.2)

**Comment by devinbinnie** [comment]
@biva Does this also happen when you open the app from a notification? I get that bug too and I've never had anyone else reproduce it before. The app just freezes and then it works once you click on it or drag it around right?

**Comment by biva** [comment]
I disabled all notifications, so I can't say :/ For me, it works again after a few seconds, I didn't notice anything about draging it around right, but I'll check.

Since when do you have this bug? Only since v5? (I didn't have it before this latest version)

**Comment by devinbinnie** [comment]
Yeah it started around v5 for me. No one else has been able to reproduce so I assumed it was issue with my machine.

**Comment by lostlont** [comment]
Same here, did not have this issue in v4.x, started in v5.0, still happens in v5.0.1. It looks like it only happens when the app is opened from the systray. I've been using it in minimized mode for a few days and this way it seems to be not happening. Also it's quite easy to reproduce it by closing and opening the systray icon. If I do the same with a minimized window then it works properly.

**Comment by biva** [comment]
Haaaa, thanks for your report: @devinbinnie won't feel alone anymore 🤣

**Comment by devinbinnie** [comment]
@lostlont Does it happen when you open the app from a notification?

**Comment by lostlont** [comment]
It's hard to tell as I have notifications enabled but I rarely click on them instantly. I'll experiment with it and get back if I see anything.

**Comment by lostlont** [comment]
Confirmed, occasionally it happens with notifications too. In this case when the window open I see my last open channel, and not the channel or person from which the notification came. Minimising and reopening the window fixes it again.

**Comment by biva** [comment]
Hello, I have also noticed that before 6.0, launching `mattermost.exe` was _reviving_ the desktop app from tray (I don't know how to say that: the application was displayed as if I click on the tray icon). It also worked when I clicked on a shortcut or on the Mattermost icon in the Windows start menu.

This doesn't work anymore: I have to click on the tray icon to revive the app. 

As both issues appeared with 6.0 and have to do with the tray icon, could they be related? Or should I create a separate issue? Can you reproduce that?

**Comment by devinbinnie** [comment]
@biva There was a fix made in v5.0.2 for the tray icon, can you try that version?

**Comment by biva** [comment]
Indeed! This works, thanks @devinbinnie !! :) I'll share here if I see improvements on the freeze issue as well with this new version.

**Comment by devinbinnie** [comment]
> After @devinbinnie 's message ([#1851 (comment)](https://github.com/mattermost/desktop/issues/1851#issuecomment-960941094)) about dragging the app around, I tested a potential workaround: instead of having the app full screen, I use it in a medium-size (it uses part of the screen only). Up to now, no freeze anymore, I'm still testing.
> 
> Does that work for you too? Would it help to find the reason behind the freeze?

I've had it happen regardless of window size unfortunately :(

**Comment by biva** [comment]
On my side it continues to work: before using this trick, it was freezing almost every time I revived it from the tray, now it never freezes anymore (only 12 hours of experience though...). Are you also using 5.0.2 @devinbinnie ?

**Comment by GuiGuiSoft** [comment]
Hello,
I have the same problem but another workaround.
When I click on the tray icon and Mattermost freeze, then I **right**-click on the tray icon and Mattermost start to respond.
If I clicked on channels, I entered something on the keyboard, all is done after the right-click.
Regards

**Comment by biva** [comment]
Hahahaha, that's a fun one, congrats for having found it :D

**Comment by devinbinnie** [comment]
> > After @devinbinnie 's message ([#1851 (comment)](https://github.com/mattermost/desktop/issues/1851#issuecomment-960941094)) about dragging the app around, I tested a potential workaround: instead of having the app full screen, I use it in a medium-size (it uses part of the screen only). Up to now, no freeze anymore, I'm still testing.
> > Does that work for you too? Would it help to find the reason behind the freeze?
> 
> I've had it happen regardless of window size unfortunately :(

I am on v5.0.2, and funny enough I played with it a bit this morning and couldn't reproduce it anymore, regardless of window size :S

**Comment by nub999** [comment]
Same issue with 5.0.2 version. app randomly hangs on startup from notification or tray.

**Comment by biva** [comment]
@devinbinnie , in fact, the bug came back for me, even with the medium-size window. I have the feeling that the app is getting tired at some point, and needs a kind of refresh.

For example, after a computer restart, the trick works for several hours, and then doesn't work anymore. At this point, if I maximize the window (full screen), and go back to medium-size, the trick works again for a few hours.

Weird, isn't it?

**Comment by devinbinnie** [comment]
@biva It is an odd bug, I'm wondering if it has to do with interaction with some other application, not sure what though. Funny enough it hasn't happened for me in a while, curious are you on the latest version of Windows 10?

**Comment by biva** [comment]
My windows 10 version is 10.0.19043 Build 19043.
This problem never happened before v6, and I had the same applications on Windows. I guess there are a lot of changes between v5 and v6, but would it be possible to investigate what could be the cause by analysing these changes?

**Comment by devinbinnie** [comment]
Ahh okay, I'm on build 19044 which I think is pretty recent. It also never happened to me before v5 but I'm not sure what could be causing it other than maybe an Electron upgrade?

**Comment by dfun90** [comment]
Can confirm this bug happening for me as well, since a while now.
Running Windows 10 Pro 21H1 (19043.1348) on my work computer, which is also connected to an Active Directory Domain.

When exactly the bug started to occur I can not say for certain, but I think sometime around the Mattermost Desktop App v4,x. Since then I did several Mattermost Server updates (Currently on 6.0.2) and also Mattermost Desktop App Updates (Currently using 5.0.2).

Oh also, I just remembered that the bug also occurs for me on my laptop at home which runs Windows 10 Pro 21H1 (N-Version... I really need to reinstall this device and not use that version) and using the latest Desktop App as well.

On both devices I have other Electron based apps running at the same time, but I doubt that this would make this issue happen. I also wonder if maybe some update to Electron could have caused this?

**Comment by Georg311** [comment]
> 

Can confirm, as soon as i right click the tray icon everything gets unstuck

Win 20H2 here

**Comment by devinbinnie** [comment]
Thanks for chiming in everyone. It stopped happening for me weeks ago, and I haven't seen it happen since. If anyone's able to reproduce while developing, I would love some help looking into this :)

**Comment by biva** [comment]
@devinbinnie it's still happening to me and my colleagues. Could you please specify how we can help investigate? How to share logs? Something else? What do you mean by _reproduce while developing? Thanks!

**Comment by JaxonWright** [comment]
Since the v6.2.1 update, Mattermost has frozen for me when clicking on desktop notifications. I am on Windows 10.

**Comment by pearj** [comment]
I noticed this since 5.0.1 I think, we're also using v6.2.1 server.

I found a workaround to unfreeze it, if you maximize or restore, it seems to get the app unstuck.

I have also notice that while frozen you can actually type, because when the app became unstuck the characters I typed while frozen were actually there.

**Comment by pearj** [comment]
> If anyone's able to reproduce while developing, I would love some help looking into this :)

@devinbinnie Do you mean to build a local copy of the desktop client and run it from VSCode with a debugger? Or do you mean attaching a Chrome Inspector and taking heap dumps or JavaScript CPU profiles?

I managed to attach a chrome inspector to mattermost by running: `Mattermost.exe --debug --inspect=5858`

Then open chrome go to `chrome://inspect/#devices`. Click configure and make sure `localhost:5858` is in the list.

Then `electron/js2c/browser_init` appears on the inspector page, click the `inspect` link and `DevTools` opens up. Although of course once I had this setup I didn't get very many messages on mattermost and it didn't freeze today.

Do you have any tips of what I should do when a freeze happens? Should I take a CPU profile and a heap dump when it freezes?

**Comment by devinbinnie** [comment]
@pearj If you're able to reproduce it in development mode, so just built locally that would be helpful.

I'm doubtful that it's going to be a renderer process issue, though I could be incorrect. Unfortunately I don't have any indication of what it could be at this point, so anything that you can find might be helpful.

Thanks for your help in advance :)

**Comment by pearj** [comment]
Hi @devinbinnie,

So I've had the inspector attached for a few weeks, and it turns out that whenever it freezes I switch to the inspector and that fixes it. Which is pretty annoying, because it means I can't start the CPU profiler when it is frozen. I suppose I could leave profiling running for a few hours and catch it that way maybe.

To me it seems like the issue is related to repainting somehow. Because once when unlocking my pc the desktop client was just solid grey, except for the top bar that contains the menus and maximise/minimise/close buttons. But I was able to type and when it unfroze and the client repainted my text was there waiting.

So I think switching to the Inspector must trigger a repaint, if it is the renderer freezing, do you have any tips about to debug that?

**Comment by devinbinnie** [comment]
@pearj Thanks for investigating :)

I'm wondering if at this point the issue is related to Chromium/Electron, as we don't fully control the rendering here. I would definitely try seeing if you can run the profiler for a long period and see if that reveals anything, otherwise it's possible that it could be fixed with an Electron upgrade? When I last checked I hadn't noticed any issues posted with them related to this, but that might be because it's so hard to debug/doesn't reproduce for everyone.

**Comment by biva** [comment]
Hello, I upgraded to 5.0.4 and the issue is still there. I also upgraded Evernote from the old version (maybe 7?, I'm not sure) to the totally different version 10.30.6, and I'm experiencing similar freezes. It might be a totally different issue, but maybe it helps understand it?

**Comment by devinbinnie** [comment]
Thanks @biva! Anything helps really :)

Looks like both issue you mentioned are both fixed in Electron v14 already, so I don't think that's the case. And for the main window, we actually aren't using `nativeWindowOpen: true` anyways, and I believe it's still defaulted to `false` in v14.

Really appreciate the investigation btw, if you have any other suggestions please let us know.

**Comment by maxmukhin88** [comment]
happens all the time. i've noticed it's likelier when CPU/RAM usage on my PC is high. i had to switch to browser window due to this annoying bug. TS, thanks for posting this

**Comment by biva** [comment]
Hello! I don't know which other apps might use Electron on my PC, but maybe Evernote? WhatsApp? How can I know which apps are using it?

Anyway, I removed Evernote and I don't use WhatsApp on the PC anymore, I'll see if it has an impact. If anybody knows other app, or can do similar test, maybe it can help understanding the source of the issue?

**Comment by biva** [comment]
Hello, a few more research:
1. https://stackoverflow.com/questions/56083523/electron-js-app-freezes-after-an-hour-unused
> the problem was solved with just one line code in the background.js file of electron, the problem came from Hardware Acceleration and because of that the app use more resources and some hours make app not respond
`app.disableHardwareAcceleration();`
by adding above code in the background process file the app runs and work very fast
2. Try to launch with `C:\Users\Bob\AppData\Local\Programs\mattermost-desktop\mattermost.exe --disable-gpu`

**Comment by tschlue** [comment]
> Hello, a few more research:
> 
>  [...]
> 
>     2. Try to launch with `C:\Users\Bob\AppData\Local\Programs\mattermost-desktop\mattermost.exe --disable-gpu`
> 
>  [...]
>

Hi, 

this works on my side. No freeze today with this way to start.

Thanks for working this out.

**Comment by Georg311** [comment]
> 2\. Try to launch with `C:\Users\Bob\AppData\Local\Programs\mattermost-desktop\mattermost.exe --disable-gpu`

This fixes the stutters for me as well

**Comment by pearj** [comment]
> > 2. Try to launch with `C:\Users\Bob\AppData\Local\Programs\mattermost-desktop\mattermost.exe --disable-gpu`
> 
> This fixes the stutters for me as well

This doesn't help for me. I tried disabling hardware acceleration via the UI and with `--disable-gpu`, neither made a difference.

**Comment by biva** [comment]
On my side, this didn't help a lot unfortunately: I still have a lot of freezes :( A solution should exist though because I use so many apps based on Electron (Zettlr, Teams, WhatsApp, ...) and Mattermost is the only one freezing :(

@devinbinnie, is there a difference between 1, 2 and 5 in my post above? (https://github.com/mattermost/desktop/issues/1851#issuecomment-1047909409). Could you please share the status of `app.disableHardwareAcceleration();` in the code of Mattermost?

**Comment by pearj** [comment]
As an aside I have discovered another interesting thing. I use https://web.whatsapp.com/ in Google Chrome, and if I click a notification I sometimes get similar freezing behaviour that I get from mattermost desktop. 

My issue generally happens when I click a notification, if I ignore the notification and open mattermost by hand it usually behaves better.


@devinbinnie I think you might be right about it being a chromium/electron issue given my chrome issues above.

> @pearj Thanks for investigating :)
> 
> I'm wondering if at this point the issue is related to Chromium/Electron, as we don't fully control the rendering here. I would definitely try seeing if you can run the profiler for a long period and see if that reveals anything, otherwise it's possible that it could be fixed with an Electron upgrade? When I last checked I hadn't noticed any issues posted with them related to this, but that might be because it's so hard to debug/doesn't reproduce for everyone.

The problem I have with running the profiler for too long is that I forget about it, and then it uses too much memory and it ends up killing all my chrome browsers! But the freezing has been a lot more frequent lately, so maybe I'll try profiling again.

**Comment by fmp777** [comment]
I've been having the same problem, but on a new Mac M1 - both a first edition M1 Mac Mini & a brand new latest Macbook Pro M1 Max.  I switched to M1 around the same time v5 of the desktop app came out i think, so I can't tell if its an issue dealing with M1 architecture or an issue of V5 or both.

For me, the app always works fine on first launch of it. Even though it takes a good minute to initialize (UI displays but is unresponsive).  Its ALWAYS locked up by the next morning after sitting all night.  The UI is completely unresponsive and it has not received any of the latest chats.  Sometimes this happens as I'm using it, like as I'm typing.

The only fix has been to use the menu item to quit the app.  Cmd+Q doesn't even work to quit the app.  Other staff using older Intel macs have not reported the same problem.

**Comment by biva** [comment]
Hello, I just changed my computer, now a Lenovo Yoga with AMD Ryzen 7 5800U, brand new installed on Windows 11, still the same issue, even if it's less frequent :(

**Comment by fmp777** [comment]
Some more info from the Mac M1. Attached is a process sample of the Renderer that is causing the freeze.  

While frozen here's what we see.
![image](https://user-images.githubusercontent.com/2659615/165111269-bf506ad5-93c9-4697-ba12-3a2222753a20.png)
![image](https://user-images.githubusercontent.com/2659615/165111286-37def7e2-a5de-466d-957f-88ca216b773b.png)

[Sample of Mattermost Helper (Renderer).txt](https://github.com/mattermost/desktop/files/8555492/Sample.of.Mattermost.Helper.Renderer.txt)


I have never had the GPU Acceleration preference enabled.

**Comment by biva** [comment]
Wow, thanks a lot @fmp777 ! Do you have any idea about what is this Mattermost Helper or a way to disable it? (even for test purpose) @devinbinnie any idea?

**Comment by fmp777** [comment]
It's the spawned process to render the UI.  When you kill the process, the whole UI inside the app window disappears. It is a normal part of functioning, whats not normal is that it hangs 100% cpu and eats so much ram.

This is normal operations:
![image](https://user-images.githubusercontent.com/2659615/165118631-ce0dfd3e-b9a6-46d9-9b1f-b5baf8767aa8.png)

**Comment by KnuX** [comment]
> I'm wondering if at this point the issue is related to Chromium/Electron

I have this issue with Mattermost (I'm a new user) but also some apps electrons likes Signal and Skred Messenger. But only via a remote desktop session (windows 11). I closed the rdp client on evening, and on the morning some apps are freezed but are still working this notifications are displayed. Maybe it's related...

**Comment by pearj** [comment]
Yeah I’m inside a Citrix session, so basically the same as rdp, and I see freezing still across Mattermost desktop client and WhatsApp running in Google chrome.  I might try disabling some of the Citrix hooks and see if I get an improvement.

**Comment by anx-ag** [comment]
What desktop app version are you running, @pearj ? There have been updates with regards to how the electron app behaves on terminal servers and 5.2.2 already has the fix builtin.

**Comment by pearj** [comment]
Whoops didn't notice I missed the message from @anx-ag, I am on v5.3.1, I'm pretty sure it's still an issue as I still see it in WhatsApp in Chrome, so I'm just assuming it's not possible to fix it on the mattermost side.

But I'll pay more attention and see if it is still actually happening, as I just got used to right-clicking to make the screen repaint.

I'm pretty certain it was still happening since January 27 though.

**Comment by pearj** [comment]
About an hour after I posted this the v5.3.1 mattermost desktop client froze until I right-clicked. So it's definitely still a thing.

**Comment by aaastafyev** [comment]
I confirm the issue.

**Comment by dlind-work** [comment]
I confirm this issue.

Windows 10, Mattermost 5.3.1, regular installation.
